doc/specs/stdlib_sparse.md

```diff
@@ -0,0 +1,305 @@
+---
+title: sparse
+---
+
+# The `stdlib_sparse` module
+
+[TOC]
+
+## Introduction
+
+The `stdlib_sparse` module provides derived types for standard sparse matrix data structures. It also provides math kernels such as sparse matrix-vector product and conversion between matrix types.
+
+## Sparse matrix derived types
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### The `sparse_type` abstract derived type
+#### Status
+
+Experimental
+
+#### Description
+The parent `sparse_type` is as an abstract derived type holding the basic common meta data needed to define a sparse matrix, as well as shared APIs. All sparse matrix flavors are extended from the `sparse_type`.
+
+```Fortran
+type, public, abstract :: sparse_type
+    integer :: nrows   !! number of rows
+    integer :: ncols   !> number of columns
+    integer :: nnz     !> number of non-zero values
+    integer :: storage !> assumed storage symmetry
+end type
```

Comment:
Just a quick thought: it might be of interest to have two DTs with different kinds for the integers, e.g.,
```fortran
type,...... :: sparse_int32_type
 ....
 integer(int32) :: nnz
 ...
end type
type,...... :: sparse_int64_type
 ....
 integer(int64) :: nnz
 ...
end type
````

Reply:
I'm not sure how hard it would be to accomplish this. Maybe in the current PR it would be enough to generalize the index `kind` from default `integer` to `integer(ik)`, so that later this extension will be easier. 

What you're suggesting is probably that there should be two `abstract` sparse types, one per length `kind`: 

```Fortran
type, public, abstract :: sparse_type

end type
#:for ik,it in INT_KINDS_TYPES
type, public, abstract, extends(sparse_type) :: sparse_${ik}$_type
    ${it}$ :: nrows   !! number of rows
    ${it}$ :: ncols   !> number of columns
    ${it}$ :: nnz     !> number of non-zero values
    ${it}$ :: storage !> assumed storage symmetry
end type sparse_${ik}$_type
!:endfor
```

Reply:
Indeed, I had something similar to your proposition in mind. However, I agree that it should be for a following PR.

Reply:
I get the feeling that adding this `COO_${s1}$_type` plus then the different integer sizes would end up with declarations such as:
```Fortran
type(CSR_int32_dp_type) :: mymatrix
```
wich I find very verbose. The target design I was striving for is to have these two things being syntactically close from a hierarchical point of view:
```Fortran
real(dp), allocatable :: dense(:,:)
type(CSR_dp) :: sparse
```
Is there a possibility to find a middle ground?

Reply:
In https://github.com/fortran-lang/stdlib/pull/760/commits/6ae038bd54ab9927aec20bb0787ba89fa5769c17 I added a `ilp` parameter following the linalg module to enable recompiling using int64. Rightnow it is set to int32, we could add a preprocessing macro to change it.

Reply:
> possibility to find a middle ground

I agree with you @jalvesz. @jvdp1's idea is correct, but maybe, who needs two index types when you can have the largest one already? In other words, could we just require that the index type for sparse matrices be a 64-bit one and define  

```
integer, parameter :: index_t = int64
```

To support 32-bit index types, one idea could be to just define the interface on the abstract type: 

```
type, abstract :: sparse
 ...
contains
   procedure(get_elem), deferred, abstract :: get_int64
   procedure :: get_int32 ! implemented, just calls get_int64
   generic :: get => get_int32, get_int64
end type
...
elemental real(dp) function get_int32(this,i,j)
    class(sparse), intent(in) :: this
    integer(int32), intent(in) :: i,j
    get_int32 = this%get(int(i,int64),int(j,int64))
end function
```

Reply:
>In other words, could we just require that the index type for sparse matrices be always a 64-bit

I'm very hesitant about this. given that with int32 one can have `nnz = 2,147,483,647` this means that, for a COO matrix of double precision values of such size, one would have 2147483647 *( 2 * 4 + 8 ) / (1024)^3 = 23 Gb of memory. Transforming the index arrays to int64 => 2147483647 *( 2 * 8 + 8 ) / (1024)^3 = 39 Gb ... And all the loops will be done by searching on larger sized indexing arrays, thus plummetting performance (do not know to which extent). Also, MKL uses int32 for the kind definition of index arrays. Shall we set a default to int64 this would make it more complicated to eventually bind against it.

Thinking about this makes me more inclined for either allowing recompilation by changing the `ilp` parameter, or indeed adding support for both sizes with separate kinds. I would in that case let the int32 with a simplified syntax and the int64 one with something like `long` added as a suffix.

Reply:
I agree with you @jalvesz: at this stage, probably the best thing to do is just to generalize the integer kind: 

> from default `integer` to `integer(ik)`

so that the implementation already supports any integer `kind`s, but a more thoughtful decision about what should be used and how flexible should that be, is deferred.

Reply:
> could we just require that the index type for sparse matrices be always a 64-bit one

The default kind should be `integer(int32)`, at least for linking them or using other libraries (like MKL Sparse BLAS).

At this stage, using `integer(ik)` is fine. There is enough WIP without this feature.